master/static/srv/ship_logs.py

```diff
@@ -581,65 +573,111 @@ def signal_passthru(signum: Any, frame: Any):
         shipper_started = True
 
-        exit_code = None
-        deadline: Optional[float] = None
         # Expect 4 messages on the doneq, one for each thread we started.
-        for _ in range(4):
+        #
+        # Ideally, we expect all four threads (two Collectors, one Shipper, and one Waiter) to
+        # finish nicely around the same time.  But there are two problems we must deal with:
+        #
+        #   - The Orphaned Grandchild Problem
+        #
+        #     If the user code starts and orphans a child process with access to our stdout and
+        #     stderr, then our Collectors can potentially be held open an indefinite amount of time
+        #     after the main subprocess exits (and the Waiter thread finishes).  The way docker
+        #     normally treats this case is that when the main process exits, every process in the
+        #     container namespace is killed immediately.  Our analog of that behavior is: after the
+        #     main process exits, we do not care about any logs from any orphaned grandchildren.
+        #
+        #     Therefore, after the main process exits, we wait up to 1 second for the Collector
+        #     threads to finish before we decide to ignore them completely.  The only reason we even
+        #     wait 1 second is to be sure not to drop logs from a well-behaved main subprocess.
+        #
+        #   - The Unreachable Master Problem
+        #
+        #     It is possible the master is temporarily not receiving logs for some reason.  This
+        #     would cause the Shipper thread to not finish promptly.  This is not the fault of user
+        #     code, and we want to try hard to preserve all logs from user code, so we wait a
+        #     total of DET_LOG_WAIT_TIME after the Waiter finishes for the shipper to keep trying.
+        exit_code = None
+        stdout_done = False
+        stderr_done = False
+        shipper_done = False
+        collector_deadline: Optional[float] = None
+        shipper_deadline: Optional[float] = None
+        while exit_code is None or not stdout_done or not stderr_done or not shipper_done:
             # Wait for an event, possibly with a deadline (if the child process already exited).
             try:
-                timeout = None if deadline is None else deadline - time.time()
-                if timeout is not None and timeout <= 0:
-                    raise queue.Empty()
+                if exit_code is None:
+                    timeout = None
+                else:
+                    # Pick the shortest deadline and calculate a timeout.
+                    assert shipper_deadline
+                    deadline = shipper_deadline
+                    if collector_deadline is not None:
+                        deadline = min(collector_deadline, shipper_deadline)
+                    timeout = deadline - time.time()
+                    if timeout <= 0:
+                        raise queue.Empty()
```

Comment:
random question: how come an exception is raised and caught later instead of having L634-652 stuff right here?

Reply:
actually, the main source of `queue.Empty()` is `doneq.get()`.  But it happens that a negative or zero timeout will cause `doneq.get(timeout)` to block indefinitely.  So if that arises during normal operation (rare but possible) we manually raise a queue.Empty() to jump into the timeout codepath.

Reply:
actually, this is a good example of why I don't actually like exception control flows (as a language feature).  It's just too non-obvious for my tastes.

But alas, we can't all be writing C code.